### PR TITLE
Removed yolo String usage for remote write; Renamed yolo Lables to ZLabels; Added further benchmarks.

### PR DESCRIPTION
**Pull request by bwplotka**
Fixes: https://github.com/thanos-io/thanos/issues/3265

**Reason for leak**: It was not leaking (: It was just over allocating a lot for every new series that comes, causing overall more memory used then usual. This is the consequence of using yolo string (no copy for string during deserialize). When we do so, we reuse some of bytes buffered in order to parse full protobuf. When we keep just labels (e.g in TSDB for new series) we actually started to keep full protobuf bytes, not just what is used for labels string. I checked and it's not worth to copy manually when needed. Not much benefits seen on benchmarks and more complex code, so removed ZLabel usage from remote write. 

(Base: nocpy labels vs normal unmarshal)
```
benchstat -delta-test=none _dev/diffd.out _dev/diffe.out
name                                                                                   old time/op    new time/op    delta
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/OK-12                         19.5µs ± 0%    20.2µs ± 0%   +3.44%
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/conflict_errors-12            27.2µs ± 0%    26.1µs ± 0%   -4.29%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/OK-12                        25.4ms ± 0%    27.5ms ± 0%   +8.09%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/conflict_errors-12           60.2ms ± 0%    56.3ms ± 0%   -6.58%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/OK-12                           25.9µs ± 0%    22.0µs ± 0%  -14.97%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/conflict_errors-12              28.3µs ± 0%    28.4µs ± 0%   +0.32%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/OK-12                          25.2ms ± 0%    27.3ms ± 0%   +8.22%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/conflict_errors-12             51.9ms ± 0%    61.3ms ± 0%  +18.05%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/OK-12                 14.3ms ± 0%    14.5ms ± 0%   +1.29%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/conflict_errors-12    12.2ms ± 0%    13.1ms ± 0%   +6.85%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_2MB_samples/OK-12                   44.5ms ± 0%    42.6ms ± 0%   -4.34%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_2MB_samples/conflict_errors-12      64.7ms ± 0%    66.3ms ± 0%   +2.37%

name                                                                                   old alloc/op   new alloc/op   delta
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/OK-12                         6.65kB ± 0%    6.33kB ± 0%   -4.84%
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/conflict_errors-12            9.66kB ± 0%    9.32kB ± 0%   -3.44%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/OK-12                        15.5MB ± 0%    13.8MB ± 0%  -10.84%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/conflict_errors-12           45.4MB ± 0%    45.4MB ± 0%   -0.00%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/OK-12                           7.05kB ± 0%    6.73kB ± 0%   -4.57%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/conflict_errors-12              10.1kB ± 0%     9.8kB ± 0%   -3.05%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/OK-12                          14.1MB ± 0%    13.5MB ± 0%   -3.81%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/conflict_errors-12             45.4MB ± 0%    45.4MB ± 0%   -0.01%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/OK-12                 22.5MB ± 0%    22.5MB ± 0%   +0.04%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/conflict_errors-12    22.5MB ± 0%    22.5MB ± 0%   +0.00%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_2MB_samples/OK-12                   42.8MB ± 0%    43.9MB ± 0%   +2.53%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_2MB_samples/conflict_errors-12      67.8MB ± 0%    67.8MB ± 0%   +0.00%

name                                                                                   old allocs/op  new allocs/op  delta
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/OK-12                           68.0 ± 0%      67.0 ± 0%   -1.47%
HandlerReceiveHTTP/typical_labels_under_1KB,_single_sample/conflict_errors-12               111 ± 0%       110 ± 0%   -0.90%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/OK-12                         5.14k ± 0%     5.14k ± 0%   -0.10%
HandlerReceiveHTTP/typical_labels_under_1KB,_2MB_of_samples/conflict_errors-12             500k ± 0%      500k ± 0%   -0.00%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/OK-12                             68.0 ± 0%      67.0 ± 0%   -1.47%
HandlerReceiveHTTP/bigger_labels_over_1KB,_single_sample/conflict_errors-12                 111 ± 0%       110 ± 0%   -0.90%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/OK-12                           5.16k ± 0%     5.14k ± 0%   -0.41%
HandlerReceiveHTTP/bigger_labels_over_1KB,_2MB_of_samples/conflict_errors-12               500k ± 0%      500k ± 0%   -0.00%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/OK-12                   47.0 ± 0%      45.0 ± 0%   -4.26%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_single_sample/conflict_errors-12      89.0 ± 0%      87.0 ± 0%   -2.25%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_2MB_samples/OK-12                    5.11k ± 0%     5.11k ± 0%    0.00%
HandlerReceiveHTTP/extremely_large_label_value_10MB,_2MB_samples/conflict_errors-12        500k ± 0%      500k ± 0%   -0.00%
```

Keeping for rest though. 

## Changes
* Optimized Hash functions for both Prometheus and Thanos (used on every write)
* Optimized conflict error write path
* Renamed Label to ZLabel for more explicit naming
* FullCopyLabel is just Label now.
* LabelSet is now ZLabelSet for explicit naming.
* Added benchmarks.

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by brancz** [comment]
Let us know when/if this is ready for a review :)

**Comment by bwplotka** [comment]
Ready for review. I decided to change naming which increased size of this PR a bit...

File: pkg/receive/writer.go [hunk 1]
```diff
@@ -61,29 +62,29 @@ func (r *Writer) Write(ctx context.Context, tenantID string, wreq *prompb.WriteR
...
-		}
+		lset := labelpb.LabelsToPromLabels(t.Labels)
 
+		var ref uint64
 		// Append as many valid samples as possible, but keep track of the errors.
-		for _, s := range t.Samples {
-			_, err = app.Add(lset, s.Timestamp, s.Value)
+		for i, s := range t.Samples {
+			if i == 0 {
+				ref, err = app.Add(lset, s.Timestamp, s.Value)
```
Comment (bwplotka) [inline pkg/receive/writer.go:71]: Major improvement here.

File: pkg/receive/hashring.go [hunk 2]
```diff
@@ -84,7 +65,11 @@ func (s simpleHashring) GetN(tenant string, ts *prompb.TimeSeries, n uint64) (st
 	if n >= uint64(len(s)) {
 		return "", &insufficientNodesError{have: uint64(len(s)), want: n + 1}
 	}
-	return s[(hash(tenant, ts)+n)%uint64(len(s))], nil
+
+	// TODO(bwplotka): Labels should be sorted already, consider removing this.
+	sort.Slice(ts.Labels, func(i, j int) bool { return ts.Labels[i].Name < ts.Labels[j].Name })
+
+	return s[(labelpb.HashWithPrefix(tenant, ts.Labels)+n)%uint64(len(s))], nil
```
Comment (bwplotka) [inline pkg/receive/hashring.go:72]: New optimized function for hashing.

File: pkg/store/labelpb/label.go [hunk 3]
```diff
@@ -13,78 +13,74 @@ import (
...
-		copy(data[i:], m.Name)
-		i = encodeVarintTypes(data, i, uint64(len(m.Name)))
-		i--
-		data[i] = 0xa
-	}
-	return len(data) - i, nil
-}
-
-func (m *Label) Unmarshal(data []byte) error {
+// Unmarshal unmarshalls gRPC protobuf into ZLabel struct. ZLabel string is directly using bytes passed in `data`.
```
Comment (bwplotka) [inline pkg/store/labelpb/label.go:79]: Noted the behaviour which can be surprising.

**Comment by bwplotka** [comment]
The "leak" removal can be seen on remote write benchmark profiles of inuse space in correct moment ;p 

Before:
image

After (note, no snappy):

**Comment by bwplotka** [comment]
Since it's a big change with extra optimizations, let's fix this against master.

Will fix release with quicker fix disabling all yolo strings for now.

**Comment by bwplotka** [comment]
Most likely to split as well

**Pull request closed without merging**